### PR TITLE
feat(plugins,toolbars,utils,controller): 添加区块内容到template

### PR DESCRIPTION
**Pull request by spongebody**
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
	- Introduced enhanced template management functionality across various components.
	- Added functions for saving and updating templates with user feedback mechanisms.
	- Implemented context-aware saving mechanisms for templates and blocks.

- **Bug Fixes**
	- Improved handling of unsaved changes when switching between templates and other elements.

- **Documentation**
	- Updated comments and naming conventions for clarity in template-related functionalities.

- **Style**
	- Enhanced code readability through formatting adjustments and logical restructuring.

- **Tests**
	- Expanded test coverage for new template management functionalities and user interactions. 

- **Chores**
	- General code clean-up and adherence to coding standards.

### CODE REVIEW

**Review by stwflyfox — APPROVED** [review]

File: packages/controller/src/useBlock.js [hunk 1]
```diff
@@ -289,7 +289,7 @@ const initBlock = async (block = {}, _langs = {}, isEdit) => {
     block.occupier = useEditorInfo().userInfo
     useLayout().layoutState.pageStatus = getCanvasStatus(block.occupier)
     addBlock(block)
-    setSaved(false)
+    !isTemplate ? setSaved(false) : setTemplateSaved(false)
```
Comment (hexqi) [inline packages/controller/src/useBlock.js]: 前面增加取反操作符会增加理解成本，建议去掉取反，调整后面的顺序：
`isTemplate ? setTemplateSaved(false): setSaved(false)`

File: packages/controller/src/useCanvas.js [hunk 2]
```diff
@@ -50,6 +50,14 @@ const defaultSchema = {
   inputs: [],
   outputs: []
 }
+const defaultTemplateState = {
+  ...defaultPageState
+}
+
+const defaultTemplateSchema = {
```
Comment (hexqi) [inline packages/controller/src/useCanvas.js]: useCanvas中模板相关的逻辑建议抽取到一个单独的文件中，使逻辑更清晰

File: packages/plugins/materials/src/block/js/usePanel.js [hunk 3]
```diff
@@ -60,4 +60,4 @@ const generateSetVisibleMethod = (usePanel) => {
 
 export const setBlockPanelVisible = generateSetVisibleMethod(useGroupPanel)
 
-export const setBlockVersionPanelVisible = generateSetVisibleMethod(useVersionSelectPanel)
+export const setBlockVersionPanelVisible = generateSetVisibleMethod(useVersionSelectPanel)
```
Comment (hexqi) [inline packages/plugins/materials/src/block/js/usePanel.js]: 文件末尾要有新行

**Comment by hexqi** [comment]
代码提交的一些规范，希望能够遵守：
- 提交代码要在PR描述中介绍清楚PR做了哪些改动
- 新的特性，特别是有UI变化的，要贴上截图或者动图让其他人更清晰了解有哪些改动
- 控制PR的大小和范围，一个PR做的事情应该是独立单一的，尽量不要在一个PR中做过多的事情，比如修复bug和新的特性应该分开提交

**Review by stwflyfox — APPROVED** [review]

**Review by hexqi — APPROVED** [review]

**Pull request merged**